backend/packages/Upgrade/src/api/services/ScheduledJobService.ts

```diff
@@ -6,57 +6,92 @@ import { ScheduledJob, SCHEDULE_TYPE } from '../models/ScheduledJob';
-        // update experiment startOn
-        await this.experimentRepository.update({ id: experiment.id }, { startOn: null });
-        return experimentService.updateState(scheduledJob.experiment.id, EXPERIMENT_STATE.ENROLLING, systemUser);
+    return await getConnection().transaction(async (transactionalEntityManager) => {
+      try {
+        const scheduledJob = await transactionalEntityManager.getRepository(ScheduledJob).findOne(id, { relations: ['experiment'] });
+
+        const currentDate = new Date();
+        const timeDif = Math.abs(currentDate.getTime() - scheduledJob.timeStamp.getTime());
+        const fiveHoursInMS = 18000000;
+        
+        if (timeDif > fiveHoursInMS) {
+          const errorMsg =  'Time Differnce of more than 5 hours is found';
+          await transactionalEntityManager.getRepository(ScheduledJob).delete({id: scheduledJob.id});
+          throw new Error(errorMsg);
+        }
+
+        if (scheduledJob && scheduledJob.experiment) {
+          const experiment = await transactionalEntityManager.getRepository(Experiment).findOne(scheduledJob.experiment.id);
+          if (scheduledJob && experiment) {
+            const systemUser = await transactionalEntityManager.getRepository(User).findOne({ email: systemUserDoc.email });
+            const experimentService = Container.get<ExperimentService>(ExperimentService);
+            // update experiment startOn
+            await transactionalEntityManager.getRepository(Experiment).update({ id: experiment.id }, { startOn: null });
+            return experimentService.updateState(scheduledJob.experiment.id, EXPERIMENT_STATE.ENROLLING, systemUser, null, transactionalEntityManager);
+          }
+        }
+        return {};
+      } catch (error) {
+        this.log.error('Error in start experiment of schedular ', error.message);
+        return error;
       }
-    }
-    return {};
+    });
   }
 
   public async endExperiment(id: string): Promise<any> {
-    const scheduledJob = await this.scheduledJobRepository.findOne(id, { relations: ['experiment'] });
-    const experiment = await this.experimentRepository.findOne(scheduledJob.experiment.id);
-    if (scheduledJob && experiment) {
-      // get system user
-      const systemUser = await this.userRepository.findOne({ email: systemUserDoc.email });
-      const experimentService = Container.get<ExperimentService>(ExperimentService);
-      // update experiment endOn
-      await this.experimentRepository.update({ id: experiment.id }, { endOn: null });
-      return experimentService.updateState(
-        scheduledJob.experiment.id,
-        EXPERIMENT_STATE.ENROLLMENT_COMPLETE,
-        systemUser
-      );
-    }
-    return {};
+    return await getConnection().transaction(async (transactionalEntityManager) => {
+      try {
+        const scheduledJob = await transactionalEntityManager.getRepository(ScheduledJob).findOne(id, { relations: ['experiment'] });
```

Comment:
We can extract the required repository only once and then use the same instead of getting every time from the transactionalEntityManager.

Reply:
Code Updated